src/main/java/se/liu/ida/hefquin/engine/query/CypherQuery.java

```diff
@@ -0,0 +1,34 @@
+package se.liu.ida.hefquin.engine.query;
+
+import se.liu.ida.hefquin.engine.query.cypher.CypherVar;
+import se.liu.ida.hefquin.engine.query.cypher.MatchClause;
+import se.liu.ida.hefquin.engine.query.cypher.ReturnStatement;
+import se.liu.ida.hefquin.engine.query.cypher.WhereCondition;
+import se.liu.ida.hefquin.engine.query.impl.MatchCypherQuery;
+import se.liu.ida.hefquin.engine.query.impl.UnionCypherQuery;
+
+import java.util.Set;
+
+public interface CypherQuery {
+    String toString();
+
+    Set<MatchClause> getMatches();
+    Set<WhereCondition> getConditions();
+    Set<ReturnStatement> getReturnExprs();
+    Set<CypherQuery> getUnion();
+    Set<CypherQuery> getIntersect();
+
+    void addMatchClause(final MatchClause match);
+    void addConditionConjunction(final WhereCondition cond);
+    void addReturnClause(final ReturnStatement ret);
+    void addQueryToUnion(final CypherQuery q);
```

Comment:
I would like our main interfaces to represent immutable objects. Is it really necessary to be able to add elements to a `CypherQuery` object (as would be possible by using these methods)? Or can your code that assembles the queries actually create the complete `CypherQuery` object after having assembled all its elements (or, alternatively, use a constructor that produces a new `CypherQuery` object by merging two given ones)? I hope so; in which case I suggest to remove these methods from the interface.